### PR TITLE
Multi-gitlab

### LINKED ISSUE #13: Multi-gitlab
https://github.com/Scribouilli/toctoctoc/pull/12 made changes for gitlab.org support
but we want to offer the possibility of logging in from different gitlab instances

For this, we need to configure a server with the triplets origin+client_id+client_secret
They cannot be versionned in git, because they're secrets
Environment variables are an option to keep a secret

Because environment variables would scale poorly with gitlab instances, we need something else

What @Ynote and i agreed on is:
- we versioned an encrypted config file in the repo
- we use an environment variable for the decryption key

We will certainly create a simple webpage to help with the encryption

### PR DESCRIPTION
**Pull request by DavidBruant**
Fixes #13

### CODE REVIEW

**Comment by DavidBruant** [comment]
Work in progress

**Comment by DavidBruant** [comment]
Lots is happening in this PR
I think it's ready to be deployed

The encrypted configuration contains only the github config for now
What i would like to do is:
1) get a review from @Ynote 
2) deploy the current version to https://toctoctoc.lechappeebelle.team/ and make sure the github part still works (on Scribouilli) (and fix if not)
3) Create the gitlab config for gitlab.com and deploy again to be able to test it
4) from then, this PR can be merged and additional gitlab instances configs can be added afterwards

File: main.js [hunk 1]
```diff
@@ -1,60 +1,54 @@
...
+  encryptedOauthServicesConfigContent, 
+  process.env.OAUTH_SERVICES_DECRYPTION_KEY
+)
 
-if(!process.env.ORIGIN){
-  console.error(`
-    Il manque la variable d'environnement "ORIGIN".
-  `)
-  process.exit(1);
+console.log('Oauth services config (after decryption): ', oauthServicesConfigContent)
```
Comment (Ynote) [inline main.js]: Do you want to use this `console.log` in production for debbugging?

File: oauth-services-config.html [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            function UTF8ToB64(s) {
+                return btoa(unescape(encodeURIComponent(s)))
+            }
+
+            const typedArrayToBase64 = typedArray => {
+                return UTF8ToB64(typedArrayToString(typedArray))
+            }
+
+            function makeCompleteKeyString(key, initializationVector){
+                console.log(key, initializationVector)
```
Comment (Ynote) [inline oauth-services-config.html:175]: Do you still want to use this `console.log`? I have the same question for the next ones.

File: prédicats.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import './types.js'
+
+/**
+ * Adapté depuis une génération par ChatGPT à partir de la définition des types.
+ * I regret nothing
```
Comment (Ynote) [inline prédicats.js:5]: 🤣

**Review by Ynote — APPROVED** [review]
This PR looks good for me. I had to read about Base64 encryption and about `crypto.subtle` because I'm not very skilled in encryption. The code is very clear about what it should do. 

I am sooooo thrilled about this feature!! Thanks 🤗

File: main.js (hunk 1 above)
Comment (DavidBruant) [reply main.js]: Thanks for the question!
I don't know...
I think yes for now for debugging purposes and maybe after a couple of succesful deployment, we can remove it (or move it to when the JSON parsing fails, still for debugging purposes but in a less critical situation)

From a security standpoint, not logging would remove a source of potential leaks (logs)

File: oauth-services-config.html (hunk 2 above)
Comment (DavidBruant) [reply oauth-services-config.html:175]: yeah, this one should go away! Thanks for catching!

**Comment by DavidBruant** [comment]
Cette PR à cet endroit est déployée avec une config mixte github + gitlab.com

Et ça fonctionne pour ce que j'arrive à voir

Je propose d'attendre un peu pour voir, mais je pense qu'on peut la merger

Normalement, ça marchera direct en ajoutant une app oauth de git.scribouilli.org dans la config (qu'on peut faire dans une autre PR pour ne pas faire durer celle-ci inutilement)

**Pull request merged**